Marlin/src/lcd/extui/dgus/DGUSDisplay.cpp

```diff
@@ -205,17 +230,61 @@ void DGUSDisplay::ProcessRx() {
         |           Command          DataLen (in Words) */
         if (command == DGUS_CMD_READVAR) {
           const uint16_t vp = tmp[0] << 8 | tmp[1];
-          //const uint8_t dlen = tmp[2] << 1;  // Convert to Bytes. (Display works with words)
-          //DEBUG_ECHOPAIR(" vp=", vp, " dlen=", dlen);
-          DGUS_VP_Variable ramcopy;
-          if (populate_VPVar(vp, &ramcopy)) {
-            if (ramcopy.set_by_display_handler)
-              ramcopy.set_by_display_handler(ramcopy, &tmp[3]);
+
+          #if ENABLED(DWIN_CREALITY_TOUCHLCD)
+
+            if (vp == 0x14 /*PIC_Now*/) {
+              const uint16_t screen_id = tmp[3] << 8 | tmp[4];
+              // In the code below DGUSLCD_SCREEN_BOOT acts as a sentinel
+              if (screen_id == 255) {
+                // DGUS OS sometimes randomly sends 255 back as an answer. Possible buffer overrun?
+                ReadCurrentScreen(); // Request again
+              }
+              else if (displayRequest != DGUSLCD_SCREEN_BOOT && screen_id != displayRequest) {
+                // A display was requested. If the screen didn't yet switch to that display, we won't give that value back, otherwise the code gets confused.
+                // The DWIN display mostly honours the PIC_SET requests from the firmware, so after a while we may want to nudge it to the correct screen
+                DEBUG_ECHOPAIR(" Got a response on the current screen: ", screen_id);
+                DEBUG_ECHOLNPAIR(" - however, we've requested screen ", displayRequest);
+              }
+              else {
+                displayRequest = DGUSLCD_SCREEN_BOOT;
+                if (current_screen_update_callback)
+                  current_screen_update_callback(static_cast<DGUSLCD_Screens>(screen_id));
+              }
+            }
```

Comment:
What is the accompanying screen firmware?

Otherwise my suggestion would be to reintegrate this entire fork, which has been well-tested: https://github.com/CR6Community/Marlin/

Reply:
I'll get to it soon, I'm sure!
<img width="247" alt="image" src="https://user-images.githubusercontent.com/698003/120943675-000c6980-c6f6-11eb-90fe-5f464333dd33.png">